cocos/misc/lodgroup-component.ts

```diff
@@ -0,0 +1,479 @@
+}
+
+@ccclass('cc.LODGroup')
+@menu('Rendering/LOD Group')
+@executeInEditMode
+export class LODGroup extends Component {
+    /**
+     * @en Object reference point in local space, e.g. center of the bound volume for all LODs
+     */
+    @serializable
+    protected _localReferencePoint: Vec3 = new Vec3(0, 0, 0);
+
+    /**
+     * @en Object Size in local space, may be auto-calculated value from object bounding box or value from user input.
+     */
+    @serializable
+    protected _objectSize = 1;
+
+    /**
+     *@en The array of LODs
+     */
+    @serializable
+    protected _LODs: LOD[] = [];
+
+    protected _lodGroup = new scene.LODGroup();
+
+    private _eventRegistered = false;
+
+    constructor () {
+        super();
+        this._lodGroup.objectSize = this._objectSize;
+    }
+
+    set localReferencePoint (val: Vec3) {
+        this._localReferencePoint.set(val);
+        this._lodGroup.localReferencePoint = val;
+    }
+
+    get localReferencePoint () { return this._localReferencePoint.clone(); }
+
+    get lodCount () { return this._LODs.length; }
+
+    @type(Number)
+    set objectSize (val: number) {
+        this._objectSize = val;
+        this._lodGroup.objectSize = val;
+    }
+
+    get objectSize () { return this._objectSize; }
+
+    @type([LOD])
+    get LODs () {
+        return this._LODs;
+    }
+
+    set LODs (LODs: LOD[]) {
+        this._LODs = LODs;
+    }
+
+    insertLOD (index: number, screenSize: number, lod: LOD | null): LOD {
+        if (!lod) {
+            lod = new LOD();
+        }
+        lod.screenRelativeTransitionHeight = screenSize;
+        this._LODs.splice(index, 0, lod);
+        this._lodGroup.LODs.splice(index, 0, lod.lod);
+        return lod;
+    }
+
+    deleteLOD (index: number) : LOD {
+        const lod = this._LODs[index];
+        this._LODs.splice(index, 1);
+        this._lodGroup.LODs.splice(index, 1);
+        return lod;
+    }
+
+    getLOD (index: number): LOD {
+        return this._LODs[index];
+    }
+
+    setLOD (index: number, lod: LOD) {
+        this._LODs[index] = lod;
+    }
+
+    recalculateBounds () {
+        LODGroupEditorUtility.recalculateBounds(this);
+    }
+
+    resetObjectSize () {
+        LODGroupEditorUtility.resetObjectSize(this);
+    }
+
+    get lodGroup () { return this._lodGroup; }
+
+    onLoad () {
+        this._lodGroup.node = this.node;
+        if (!this._eventRegistered) {
+            // this.node.on(NodeEventType.ACTIVE_IN_HIERARCHY_CHANGED, this._removeUnneededModel, this);
+            this.node.on(NodeEventType.COMPONENT_REMOVED, this._onRemove, this);
+            this._eventRegistered = true;
+        }
+        // generate default lod for lodGroup
+        if (this.lodCount < 1) {
+            const size = _DEFAULT_SCREEN_OCCUPATION.length;
+            for (let i = 0; i < size; i++) {
+                const lod = new LOD();
+                this.insertLOD(i, _DEFAULT_SCREEN_OCCUPATION[i], lod);
```

Comment:
It seems lod is not needed. It may generates many temporary js objects doing like this? I am not sure.

Reply:
LOD can be constructed internally at the API insertLOD, if lod is undefined or null. Here we remove it